zkevm-circuits/src/evm_circuit/execution/msize.rs

```diff
@@ -0,0 +1,173 @@
+use crate::{
+    evm_circuit::{
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition, Transition::Delta,
+            },
+            from_bytes, RandomLinearCombination,
+        },
+    },
+    util::Expr,
+};
+use array_init::array_init;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct MsizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    value: RandomLinearCombination<F, 8>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for MsizeGadget<F> {
+    const NAME: &'static str = "MSIZE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::MSIZE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // memory_size is limited to 64 bits so we only consider 8 bytes
```

Comment:
I think `memory_size` is capped at 32 bits due to quadratic part of memory expansion gas cost, but I am also planning to file a PR to make this part consistent with spec, we could update this then since there are other trivial things need to be change at the same time.

Reply:
I need some advice here, since every `assign_exec_step` is called in `execution.rs` by different execution_state. For memory_size, in current step it has previous step's memory_size, but that's not we want. See code [here](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L388-L403).

Reply:
I was not aware of this behavior from geth!  What we could do is modify the geth Execution Trace to cut the memory to the proper size (by checking the observed memory size of the previous step).  This way each step in the geth trace will have the memory size before the step.  Would that solve the issue?

Reply:
I think something like this would fix the memory sizes of the geth exec steps;
```rust
fn fix_geth_trace_memory_size(trace: &mut GethExecTrace) {
    let mut mem_sizes = vec![0; trace.struct_logs.len()];
    let mut call_mem_size_stack = Vec::new();
    for i in 1..trace.struct_logs.len() {
        let step_prev = &trace.struct_logs[i - 1];
        let step = &trace.struct_logs[i];
        mem_sizes[i] = match step.depth as isize - step_prev.depth as isize {
            // Same call context
            0 => step_prev.memory.0.len(),
            // into new call context
            1 => {
                call_mem_size_stack.push(step_prev.memory.0.len());
                0
            }
            // return from call context
            -1 => call_mem_size_stack.pop().expect("call stack is empty"),
            _ => unreachable!(),
        };
    }
    for i in 0..trace.struct_logs.len() {
        trace.struct_logs[i].memory.0.truncate(mem_sizes[i]);
    }
}
```
We could have this transformation in the `GethExecTrace` deserialization step for example.  What do you think?

Reply:
yeah, using previous step will break when it returns and then use the memory size instantly. Using call stack style to fix this looks great!

Reply:
I tried https://github.com/appliedzkp/zkevm-circuits/commit/432069605b3e7eeec806eda3f313b60edc533e4a, sadly it did not solve the problem. The traced data looks good to me.

The cause is in witness.rs [step_convert](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L388-L403):
```rust
execution_state: ExecutionState::from(step), // MSIZE
...
memory_size: match prev {
    None => 0, // MSIZE
    Some(prev_step) => (prev_step.memory_size as u64) / 32,  // STOP
},
```
`MSIZE step` has 0 memory_size, and `STOP` step has proper memory_size that `MSIZE` wanted.

So I add a special case `memory_size` assignment for `MSIZE`. What do you think?

Reply:
I've taken a look and it seems it didn't work because the external_tracer works directly with `Vec<GethExecStep`, so the Deserialize on `GethExecTrace` is not called.  We can call `fix_geth_trace_memory_size` in the tracer function of the external_tracer.

I've updated the commit I shared before.  Can you try with this? https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7
I've also added a test with your example, and now the `MSTORE` has memory size at 0.

Reply:
I changed as https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7 and tried yestoday, still same.